apps/src/lab2/views/ExtraLinks.tsx

```diff
@@ -14,61 +15,86 @@ interface PermissionResponse {
+}
+
+async function fetchExtraLinksData(
+  levelId: number,
+  channelId?: string
+): Promise<ExtraLinksData> {
+  // Fetch permissions.
+  const permissionsResponse = await HttpClient.fetchJson<PermissionResponse>(
+    '/api/v1/users/current/permissions'
+  );
+  const {permissions} = permissionsResponse.value;
+
+  // Fetch level link data.
+  let levelLinkData: ExtraLinksLevelData | undefined;
+  if (
+    permissions.includes(PERMISSIONS.LEVELBUILDER) ||
+    permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)
+  ) {
+    const levelLinkDataResponse =
+      await HttpClient.fetchJson<ExtraLinksLevelData>(
+        `/levels/${levelId}/extra_links`
+      );
+    levelLinkData = levelLinkDataResponse.value;
+  }
+
+  // Fetch project link data.
+  let projectLinkData: ExtraLinksProjectData | undefined;
+  if (permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)) {
+    const levelProjectDataResponse =
+      await HttpClient.fetchJson<ExtraLinksProjectData>(
+        `/projects/${channelId}/extra_links`
+      );
+    projectLinkData = levelProjectDataResponse.value;
+  }
+
+  // Return fetched permissions and link data.
+  return {
+    levelLinkData,
+    projectLinkData,
+  };
 }
 
 // If the user has permission to see extra links, fetch extra links for the level,
 // then display a modal with the link data.
 const ExtraLinks: React.FunctionComponent<ExtraLinksProps> = ({
   levelId,
 }: ExtraLinksProps) => {
-  const {loading, data} = useFetch('/api/v1/users/current/permissions');
+  const [isLoading, setIsLoading] = useState(false);
+  const [extraLinksData, setExtraLinksData] = useState<ExtraLinksData | null>(
+    null
+  );
   const [isModalOpen, setIsModalOpen] = useState(false);
-  const [linkData, setLinkData] = useState<ExtraLinksResponse | null>(null);
 
-  useEffect(() => {
-    const permissionData = data
-      ? (data as PermissionResponse)
-      : {permissions: []};
-    if (permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER)) {
-      try {
-        HttpClient.fetchJson<ExtraLinksResponse>(
-          `/levels/${levelId}/extra_links`
-        ).then(response => {
-          setLinkData(response.value);
-        });
-      } catch (e) {
-        console.error('Error fetching extra links', e);
-      }
-    }
-  }, [data, levelId]);
-
-  const permissionData = data
-    ? (data as PermissionResponse)
-    : {permissions: []};
+  const channelId = useAppSelector(
+    state => state.lab.channel && state.lab.channel.id
+  );
 
-  if (!permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER)) {
-    return <></>;
-  }
+  useEffect(() => {
+    setIsLoading(true);
+    fetchExtraLinksData(levelId, channelId).then(data => {
+      setExtraLinksData(data);
+      setIsLoading(false);
+    });
+  }, [levelId, channelId]);
+  const {levelLinkData, projectLinkData} = extraLinksData || {};
 
-  return loading || !linkData ? null : (
+  return isLoading || (!levelLinkData && !projectLinkData) ? null : (
```

Comment:
Nice, much cleaner! One last thing, totally up to you - I personally prefer to early return rather than putting a whole react component in a ternary since it reads a little easier to me. Either way works, especially since it's not a lot of TSX. But for example I might do:
```
if (isLoading || (!levelLinkData && !projectLinkData)) {
  return null;
}

return <ExtraLinksModal ... />

Reply:
I don't really have a personal preference so can update.